uniffi_bindgen/src/bindings/gecko/templates/InterfaceTemplate.cpp

```diff
@@ -0,0 +1,97 @@
+// This file was autogenerated by some hot garbage in the `uniffi` crate.
+// Trust me, you don't want to mess with it!
+
+{% import "macros.cpp" as cpp %}
+
+#include "mozilla/dom/{{ obj.name()|class_name_webidl }}.h"
+
+namespace mozilla {
+namespace dom {
+
+// Cycle collection boilerplate for our interface implementation. `mGlobal` is
+// the only member that needs to be cycle-collected; if we ever add any JS
+// object members or other interfaces to the class, those should be collected,
+// too.
+NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE({{ obj.name()|class_name_webidl }}, mGlobal)
+NS_IMPL_CYCLE_COLLECTING_ADDREF({{ obj.name()|class_name_webidl }})
+NS_IMPL_CYCLE_COLLECTING_RELEASE({{ obj.name()|class_name_webidl }})
+NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION({{ obj.name()|class_name_webidl }})
+  NS_WRAPPERCACHE_INTERFACE_MAP_ENTRY
+  NS_INTERFACE_MAP_ENTRY(nsISupports)
+NS_INTERFACE_MAP_END
+
+{{ obj.name()|class_name_cpp }}::{{ obj.name()|class_name_cpp }}(
+  nsIGlobalObject* aGlobal,
+  int64_t aHandle
+) : mGlobal(aGlobal), mHandle(aHandle) {}
+
+{{ obj.name()|class_name_cpp }}::~{{ obj.name()|class_name_cpp }}() {
+  if (mHandle != -1) {
+    {{ obj.ffi_object_free().name() }}(mHandle);
+    mHandle = -1;
+  }
+}
+
+JSObject* {{ obj.name()|class_name_cpp }}::WrapObject(
+  JSContext* aCx,
+  JS::Handle<JSObject*> aGivenProto
+) {
+  return dom::{{ obj.name()|class_name_cpp }}_Binding::Wrap(aCx, this, aGivenProto);
+}
+
+{% for cons in obj.constructors() %}
+/* static */
+already_AddRefed<{{ obj.name()|class_name_cpp }}> {{ obj.name()|class_name_cpp }}::Constructor(
+  {% call cpp::decl_constructor_args(cons) %}
+) {
+  {%- if cons.throws().is_some() %}
+  RustError err = {0, nullptr};
+  {% endif %}
+  auto handle = {{ cons.ffi_func().name() }}(
+    {%- let args = cons.arguments() %}
+    {% call cpp::to_ffi_args(args) -%}
+    {%- if cons.throws().is_some() %}
+    {% if !args.is_empty() %},{% endif %}&err
+    {% endif %}
+  );
+  {%- if cons.throws().is_some() %}
+  if (err.mCode) {
+    aRv.ThrowOperationError(err.mMessage);
```

Comment:
Error types are fun. Firefox has a hard-coded list of DOM exceptions (there aren’t that many! And I think the guidance for WebIDL APIs is to use a limited set of built-in exceptions, like `AbortError`, `InvalidStateError`, etc.) that can be thrown on an `ErrorResult&`...which is kind of at odds with how we expose errors. We might need to use `jsapi` here to make a custom error object, and then call `aRv.ThrowJSException`.

Propagating errors with all the right type information is still a big `TODO` for this PR.

Reply:
I filed https://github.com/mozilla/uniffi-rs/issues/295.